File: lesson_18/src/main/java/org/lesson18/ConsoleRunner.java
package org.lesson18;


// /buy id_покупателя id_товара - реализовать возможность “покупки товара” по id покупателя и товара.

import org.hibernate.SessionFactory;
import org.hibernate.cfg.Configuration;

import java.util.Scanner;
import java.util.UUID;

public class ConsoleRunner {
    public static void showProductsByConsumer(SessionFactory factory, String fio) {
        try (var session = factory.getCurrentSession()) {
            session.getTransaction().begin();
            var list = session.createQuery("SELECT c FROM Customer c WHERE c.fio = :fio", Customer.class)
                    .setParameter("fio", fio)
                    .getResultList();
            if (list.isEmpty()) {
                System.out.println("Не найдено покупателей с именем " + fio);
            } else {
                list.forEach(customer -> {
                    System.out.println("Список покупок для " + fio + " c id " + customer.getId());
                    customer.getProducts().forEach(product ->
                            System.out.println(product.getTitle()));
                });
                System.out.println();
            }
            session.getTransaction().commit();
        }
    }

    public static void showConsumersByProductTitle(SessionFactory factory, String title) {
        try (var session = factory.getCurrentSession()) {
            session.getTransaction().begin();
            var list = session.createQuery("SELECT c FROM Product c WHERE c.title = :title", Product.class)
                    .setParameter("title", title)
                    .getResultList();
            if (list.isEmpty()) {
                System.out.println("Не найдено товаров с названием " + title);
            } else {
                list.forEach(product -> {
                    System.out.println("Список покупателей для " + title + " c id " + product.getId());
                    product.getCustomers().forEach(customer ->
                            System.out.println(customer.getFio()));
                });
                System.out.println();
            }
            session.getTransaction().commit();
        }
    }

    public static void deleteConsumer(SessionFactory factory, String fio) {
        try (var session = factory.getCurrentSession()) {
            session.getTransaction().begin();
            var list = session.createQuery("SELECT c FROM Customer c WHERE c.fio = :fio", Customer.class)
                    .setParameter("fio", fio)
                    .getResultList();
            list.forEach(session::delete);
            session.getTransaction().commit();
        }
    }

    public static void deleteProduct(SessionFactory factory, String title) {
        try (var session = factory.getCurrentSession()) {
            session.getTransaction().begin();
            var list = session.createQuery("SELECT c FROM Product c WHERE c.title = :title", Product.class)
                    .setParameter("title", title)
                    .getResultList();
            list.forEach(session::delete);
            session.getTransaction().commit();
        }
    }

    public static void buy(SessionFactory factory, String consumerId, String productId, String count) {
        try (var session = factory.getCurrentSession()) {
            session.getTransaction().begin();
            Product product = session.get(Product.class, UUID.fromString(productId));

            PurchaseDetails pd = new PurchaseDetails(product.getId(), UUID.fromString(consumerId), product.getPrice(), Long.valueOf(count));
            session.persist(pd);
            session.getTransaction().commit();
        }
    }

    public static void run() {
        try (var factory = new Configuration().configure("hibernate.cfg.xml").buildSessionFactory()) {

            while (true) {
                Scanner s = new Scanner(System.in);
                var args = s.nextLine().split(" ");
                switch (args[0]) {
                    case "/showProductsByConsumer" -> {
                        if (args.length != 2) {
                            System.out.println("У команды /showProductsByConsumer должен быть 1 аргумент");
                            continue;
                        }
                        showProductsByConsumer(factory, args[1]);
                    }
                    case "/showConsumersByProductTitle" -> {
                        if (args.length != 2) {
                            System.out.println("У команды /showConsumersByProductTitle должен быть 1 аргумент");
                            continue;
                        }
                        showConsumersByProductTitle(factory, args[1]);
                    }
                    case "/deleteConsumer" -> {
                        if (args.length != 2) {
                            System.out.println("У команды /deleteConsumer должен быть 1 аргумент");
                            continue;
                        }
                        deleteConsumer(factory, args[1]);
                    }
                    case "/deleteProduct" -> {
                        if (args.length != 2) {
                            System.out.println("У команды /deleteProduct должен быть 1 аргумент");
                            continue;
                        }
                        deleteProduct(factory, args[1]);
                    }
                    case "/buy" -> {
                        if (args.length != 4) {
                            System.out.println("У команды /buy должен быть 1 аргумент");
                            continue;
                        }
                        buy(factory, args[1], args[2], args[3]);
                    }
                    case "/quit" -> {
                        return;
                    }
                    default -> {
                        System.out.println("Неизвестная команда: " + args[0]);
                    }
                }

            }
        }
    }
}

File: lesson_18/src/main/kotlin/org/lesson18kt/ConsoleRunner.kt
package org.lesson18kt

import org.hibernate.SessionFactory
import org.hibernate.cfg.Configuration
import java.util.UUID


// /buy id_покупателя id_товара - реализовать возможность “покупки товара” по id покупателя и товара.

fun run() = Configuration().configure("hibernate2.cfg.xml").buildSessionFactory().use { factory ->
    while (true) {
        val args = readLine()?.split(" ")
        args?.let {
            when (args[0]) {
                "/showProductsByConsumer" -> showProductsByConsumer(factory, args)
                "/showConsumersByProductTitle" -> showConsumersByProductTitle(factory, args)
                "/deleteConsumer" -> deleteConsumer(factory, args)
                "/deleteProduct" -> deleteProduct(factory, args)
                "/buy" -> buy(factory, args)
                "/quit" -> return
                else -> println("Ошибка в команде")
            }
        }
    }
}

fun showProductsByConsumer(factory: SessionFactory, args: List<String>) = factory.currentSession.use { session ->
    val fio = args[1]

    session.transaction.begin()
    session.createQuery("SELECT c FROM Customer c WHERE c.fio = :fio", Customer::class.java)
        .setParameter("fio", fio)
        .resultList
        .forEach { customer ->
            println("Список товаров для ${customer.fio} с id = ${customer.id}")
            customer.products?.forEach { product ->
                println(product.title)
            }
        }

    session.transaction.commit()
}

fun showConsumersByProductTitle(factory: SessionFactory, args: List<String>) = factory.currentSession.use { session ->
    val title = args[1]

    session.transaction.begin()
    session.createQuery("SELECT p FROM Product p WHERE p.title = :title", Product::class.java)
        .setParameter("title", title)
        .resultList
        .forEach { product ->
            println("Список покупателей для ${product.title} с id = ${product.id}")
            product.customers?.forEach { customer ->
                println(customer.fio)
            }
        }

    session.transaction.commit()
}

fun deleteConsumer(factory: SessionFactory, args: List<String>) = factory.currentSession.use { session ->
    val fio = args[1]

    session.transaction.begin()
    session.createQuery("SELECT c FROM Customer c WHERE c.fio = :fio", Customer::class.java)
        .setParameter("fio", fio)
        .resultList
        .forEach { session.remove(it) }

    session.transaction.commit()
}

fun deleteProduct(factory: SessionFactory, args: List<String>) = factory.currentSession.use { session ->
    val title = args[1]

    session.transaction.begin()
    session.createQuery("SELECT p FROM Product p WHERE p.title = :title", Product::class.java)
        .setParameter("title", title)
        .resultList
        .forEach { session.remove(it) }

    session.transaction.commit()
}

fun buy(factory: SessionFactory, args: List<String>) = factory.currentSession.use { session ->
    val customerId = UUID.fromString(args[1])
    val productId = UUID.fromString(args[2])
    val count = args[3].toLong()

    session.transaction.begin()
    session.find(Product::class.java, productId)?.let { product ->
        session.persist(
            PurchaseDetails(
                customerId = customerId,
                productId = productId,
                price = product.price,
                count = count
            )
        )
    }

    session.transaction.commit()
}
